pkg/util/goroutinemap/common/errors.go

```diff
@@ -0,0 +1,96 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package common contains helper methods and common functionality used by
+// all RoroutineMap implementations.
+package common
+
+import "fmt"
+
+// NewAlreadyExistsError returns a new instance of AlreadyExists error.
```

Comment:
@saad-ali @smarterclayton I think we can take lines 23-54 (everything but the exponential back off stuff) and just keep it in gorroutinemap pkgand then duplicate for the new datastructure. I know the idea of duplicating this code breaks your heat. it's boiler plate not a big deal to duplicate here.